stan/math/mix/functor/wolfe_line_search.hpp

```diff
@@ -0,0 +1,1024 @@
+ * `best`.
+ *    **Ends when** a test fails, \f$\alpha>\text{opt.max_alpha}\f$, or
+ * iteration limit is reached. The function **returns** the last passing step
+ * with `Wolfe`.
+ *
+ * 3. **Expansion & right-bracketing.**
+ *    Otherwise, continue evaluating larger \f$\alpha\f$ values:
+ *    - If Armijo holds and \f$\phi'(\alpha)>0\f$, promote the left endpoint
+ *      `low ← high` and *expand* again.
+ *      (We are on the “far side” of the minimum.)
+ *    - If Armijo holds but \f$\phi'(\alpha)\le 0\f$, or if Armijo fails, a
+ * right endpoint has been found -> stop expanding and proceed to zoom.
+ *    **Ends when** a right endpoint is found,
+ * \f$\alpha>\text{opt.max_alpha}\f$, or iteration limit is reached.
+ *
+ * 4. **Safety bisection (optional).**
+ *    If both endpoints currently have positive directional derivatives, take
+ * pure bisection steps until a sign change is observed or the interval is
+ * small.
+ *
+ * 5. **Cubic/bisection zoom.**
+ *    Repeatedly propose an interior point \f$\alpha_m\f$ using a cubic
+ * interpolation of the bracket end-points (falling back to bisection when the
+ * cubic is degenerate), evaluate there, and shrink the bracket according to:
+ *    - If Armijo and curvature hold → **returns** `Wolfe`.
+ *    - If Armijo holds but curvature fails → keep the better side; prefer the
+ * sub-interval with a derivative sign change.
+ *    - If Armijo fails → move the right endpoint left.
+ *    Throughout zoom we remember the best Armijo-OK point `best` in case
+ *    strong-Wolfe cannot be satisfied.
+ *    **Ends when** the interval width is \f$\le\text{opt.min_alpha}\f$, a
+ * convergence tolerance triggers, or the iteration limit is reached.
+ *
+ * 6. **Convergence/guard rails.**
+ *    After each evaluation we check:
+ *    - Gradient tolerance: \f$|\phi'(\alpha)| \le
+ *      \max(\text{opt.abs\_grad\_threshold},\
+ * \text{opt.rel\_grad\_threshold}\,|\phi'(0)|)\f$.
+ *    - Objective change tolerance (for very small steps):
+ *      \f$|\phi(\alpha)-\phi(0)| \le
+ *      \max(\text{opt.abs\_obj\_threshold},\
+ * \text{opt.rel\_obj\_threshold}\,(1+|\phi(0)|))\f$ and \f$\alpha <
+ * \text{opt.min_alpha}\f$. If triggered, the routine returns one of
+ * `ConvergedGradient`, `ConvergedObjective`, or
+ * `ConvergedObjectiveAndGradient`. If neither test is met but the bracket
+ * becomes too small, it returns `IntervalTooSmall` (accepting the current point
+ * only if it satisfies Armijo).
+ *
+ * If zoom exits without a strong-Wolfe point but an Armijo-OK point was seen,
+ * the routine **returns** `Armijo` at that best point; otherwise it **returns**
+ * `Fail`.
+ *
+ * ### Contract for the step evaluator `UpdateFun`
+ *
+ * The line search delegates *all* numerical work to a user-supplied callable
+ * `update_fun` that evaluates the objective and its directional derivative at a
+ * trial step and prepares the candidate state for acceptance.
```

Comment:
I'm a bit confused here. So the Stan user needs to provide `UpdateFun` or is all this still happening internally?